datalad/support/dsconfig.py

```diff
@@ -0,0 +1,319 @@
+# emacs: -*- mode: python; py-indent-offset: 4; tab-width: 4; indent-tabs-mode: nil -*-
+# ex: set sts=4 ts=4 sw=4 noet:
+# ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+#
+#   See COPYING file distributed along with the datalad package for the
+#   copyright and license terms.
+#
+# ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+"""
+"""
+
+import logging
+from datalad.cmd import Runner
+import re
+import os
+from os.path import join as opj, exists
+
+lgr = logging.getLogger('datalad.support.dsconfig')
+
+cfg_kv_regex = re.compile(r'([^=]+)=(.*)')
+cfg_section_regex = re.compile(r'(.*)\.[^.]+')
+cfg_sectionoption_regex = re.compile(r'(.*)\.([^.]+)')
+
+
+def _parse_gitconfig_dump(dump, store, replace):
+    if replace:
+        # if we want to replace existing values in the store
+        # collect into a new dict and `update` the store at the
+        # end. This way we get the desired behavior of multi-value
+        # keys, but only for the current source
+        dct = {}
+    else:
+        # if we don't want to replace value, perform the multi-value
+        # preserving addition on the existing store right away
+        dct = store
+    for line in dump.split('\n'):
+        if not line:
+            continue
+        k, v = cfg_kv_regex.match(line).groups()
+        present_v = dct.get(k, None)
+        if present_v is None:
+            dct[k] = v
+        else:
+            if isinstance(present_v, tuple):
+                dct[k] = present_v + (v,)
+            else:
+                dct[k] = (present_v, v)
+    if replace:
+        store.update(dct)
+    return store
```

Comment:
just a note that replace=False is not actively used as far as I see (ran only when store is `{}` anyways)